drivers/frequency/adf4377/adf4377.c

```diff
@@ -0,0 +1,668 @@
+
+	/* Set output Amplitude */
+	ret = adf4377_update(dev, ADF4377_REG(0x19),
+			     ADF4377_CLKOUT2_OP_MSK | ADF4377_CLKOUT1_OP_MSK,
+			     ADF4377_CLKOUT1_OP(dev->clkout_op) | ADF4377_CLKOUT2_OP(dev->clkout_op));
+
+	return ret;
+}
+
+/**
+ * @brief Initializes the ADF4377.
+ * @param device - The device structure.
+ * @param init_param - The structure containing the device initial parameters.
+ * @return Returns SUCCESS in case of success or negative error code.
+ */
+int32_t adf4377_init(struct adf4377_dev **device,
+		     struct adf4377_init_param *init_param)
+{
+	int32_t ret;
+	uint8_t device_id;
+	uint8_t chip_type;
+	struct adf4377_dev *dev;
+
+	dev = (struct adf4377_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+	dev->dev_id = init_param->dev_id;
+	dev->spi4wire = init_param->spi4wire;
+	dev->clkin_freq = init_param->clkin_freq;
+	dev->cp_i = init_param->cp_i;
+	dev->muxout_default = init_param->muxout_select;
+	dev->ref_doubler_en = init_param->ref_doubler_en;
+	dev->f_clk = init_param->f_clk;
+	dev->clkout_op = init_param->clkout_op;
+
+	/* GPIO Chip Enable */
+	ret = gpio_get_optional(&dev->gpio_ce, init_param->gpio_ce_param);
+	if (ret < 0)
+		goto error_dev;
+
+	if(dev->gpio_ce) {
+		ret = gpio_direction_output(dev->gpio_ce, GPIO_HIGH);
+		if (ret < 0)
+			goto error_gpio_ce;
+	}
+
+	/* SPI */
+	ret = spi_init(&dev->spi_desc, init_param->spi_init);
+	if (ret < 0)
+		goto error_gpio_ce;
+
+	/* Software Reset */
+	ret = adf4377_soft_reset(dev);
+	if (ret < 0)
+		return ret;
+
+	ret = adf4377_spi_write(dev, ADF4377_REG(0x00),
+				ADF4377_LSB_FIRST_R(dev->spi_desc->bit_order) |
+				ADF4377_LSB_FIRST(dev->spi_desc->bit_order) |
+				ADF4377_SDO_ACTIVE_R(dev->spi4wire) |
+				ADF4377_SDO_ACTIVE(dev->spi4wire) |
+				ADF4377_ADDRESS_ASC_R(ADF4377_ADDR_ASC_AUTO_DECR) |
+				ADF4377_ADDRESS_ASC(ADF4377_ADDR_ASC_AUTO_DECR));
+	if (ret < 0)
+		return ret;
+
+	/* Read Chip Type */
+	ret = adf4377_spi_read(dev, ADF4377_REG(0x03), &chip_type);
+	if (ret < 0)
+		return ret;
+
+	if (chip_type != ADF4377_CHIP_TYPE)
+		return ret;
+
+	/* Scratchpad Check */
+	ret = adf4377_check_scratchpad(dev);
+	if (ret < 0)
+		return ret;
+
+	ret = adf4377_spi_read(dev, ADF4377_REG(0x04), &device_id);
+	if (ret < 0)
+		goto error_spi;
+
+	if (dev->dev_id == device_id) {
+		ret = gpio_get_optional(&dev->gpio_enclk1, init_param->gpio_enclk1_param);
+		if (ret < 0)
+			goto error_spi;
+
+		if(dev->gpio_enclk1) {
+			ret = gpio_direction_output(dev->gpio_enclk1, GPIO_HIGH);
+			if (ret < 0)
+				goto error_gpio_enclk1;
+		}
+
+		if (dev->dev_id == ADF4377) {
+			ret = gpio_get_optional(&dev->gpio_enclk2, init_param->gpio_enclk2_param);
+			if (ret < 0)
+				goto error_gpio_enclk1;
+
+			if(dev->gpio_enclk2) {
+				ret = gpio_direction_output(dev->gpio_enclk2, GPIO_HIGH);
+				if (ret < 0)
+					goto error_gpio_enclk2;
+			}
+		}
+	} else {
+		ret = -ENODEV;
+		goto error_spi;
+	}
+
+	ret = adf4377_setup(dev);
+	if (ret < 0)
+		goto error_spi;
```

Comment:
`goto error_gpio_enclk2;`

Reply:
this is valid only for `ADF4377`. 
For `ADF4378` is `goto error_gpio_enclk1;`
Will add extra conditions